man/man1/ils.1
.TH ILS 1 
.ad
.fi
.SH NAME
ils
\-
list inode information
.SH SYNOPSIS
.na
.nf
.ad
.fi
\fBils\fR [\fB-eorvV\fR] [\fB-f \fIfstype\fR]
\fIdevice\fR [\fIstart-stop\fR ...]

\fBils\fR [\fB-aAlLvVzZ\fR] [\fB-f \fIfstype\fR]
\fIdevice\fR [\fIstart-stop\fR ...]
.SH DESCRIPTION
.ad
.fi
\fBils\fR opens the named \fIdevice\fR and lists inode information.
By default, \fBils\fR lists only the inodes of removed files.

Arguments:
.IP \fB-e\fR
List every inode in the file system.
.IP "\fB-f\fI fstype\fR"
Specifies the file system type. The default file system type
is system dependent. With most UNIX systems the default type
is \fBffs\fR (Berkeley fast file system). With Linux the default
type is \fBext2fs\fR (second extended file system).
.IP \fB-o\fR
List only inodes of removed files that are still open or executing.
This option is short-hand notation for \fB-aL\fR
(see the \fBfine controls\fR section below).
.IP \fB-r\fR
List only inodes of removed files. This option is short-hand notation
for \fB-LZ\fR
(see the \fBfine controls\fR section below).
.IP \fB-v\fR
Turn on verbose mode, output to stderr.
.IP \fB-V\fR
Turn on verbose mode, output to stdout.
.IP \fIdevice\fR
Disk special file, or regular file containing a disk image.
On UNIX systems, raw mode disk access may give better performance
than block mode disk access.  LINUX disk device drivers support
only block mode disk access.
.IP "\fIstart-stop\fR ..."
Examine the specified inode number or number range. Either the
\fIstart\fR, the \fIstop\fR, or the \fI-stop\fR may be omitted.
.PP
Fine controls:
.IP \fB-a\fR
List only allocated inodes: these belong to files with at least one
directory entry in the file system, and to removed files that
are still open or executing.
.IP \fB-A\fR
List only unallocated inodes: these belong to files that no longer
exist.
.IP \fB-l\fR
List only inodes with at least one hard link. These belong to files
with at least one directory entry in the file system.
.IP \fB-L\fR
List only inodes without any hard links. These belong to files that no
longer exist, and to removed files that are still open or executing.
.IP \fB-z\fR
List only inodes with zero status change time. Presumably, these
inodes were never used.
.IP \fB-Z\fR
List only inodes with non-zero status change time. Presumably, these
belong to files that still exist, or that existed in the past.
.PP
The output format is in time machine format, as described in
tm-format(5). The output begins with a two-line header that
describes the data origin, and is followed by a one-line header
that lists the names of the data attributes that make up the
remainder of the output:
.IP \fBst_ino\fR
The inode number.
.IP \fBst_alloc\fR
Allocation status: `a' for allocated inode, `f' for free inode.
.IP \fBst_uid\fR
Owner user ID.
.IP \fBst_gid\fR
Owner group ID.
.IP \fBst_mtime\fR
UNIX time (seconds) of last file modification.
.IP \fBst_atime\fR
UNIX time (seconds) of last file access.
.IP \fBst_ctime\fR
UNIX time (seconds) of last inode status change.
.IP \fBst_dtime\fR
UNIX time (seconds) of file deletion (LINUX only).
.IP \fBst_mode\fR
File type and permissions (octal).
.IP \fBst_nlink\fR
Number of hard links.
.IP \fBst_size\fR
File size in bytes.
.IP \fBst_block0,st_block1\fR
The first two entries in the direct block address list.
.SH SEE ALSO
.na
.nf
mactime(1), mtime, atime, ctime reporter
tm-format(5), time machine data format
.SH BUGS
.ad
.fi
\fBils\fR should support more file system types. Right now, support
is limited to \fBext2fs\fR when built on Linux, and \fBffs\fR when
built on Solaris and BSD systems.
.SH LICENSE
.na
.nf
This software is distributed under the IBM Public License.
.SH AUTHOR(S)
.na
.nf
Wietse Venema
IBM T.J. Watson Research
P.O. Box 704
Yorktown Heights, NY 10598, USA


CHANGES
Sat Aug 29 2009

o	grave-robber did not record the initial directory in the body file.

o	Bit-rot: FreeBSD 7 support, including ``struct acct'' changes.

o	Bit-rot: lastcomm format strings.

Sun Jan 11 20:01:27 EST 2009

o	Bit-rot: OpenBSD 4.x support (still TODO: file system tools).

Thu Jun  7 10:54:44 EDT 2007

o	Bit-rot: FreeBSD 6.x support.

Mon Apr  9 09:36:31 EDT 2007

o	Bit-rot: the 64-bit workaround for LINUX lseek() is no
	longer needed, and it didn't compile on some systems.

Thu Jan 11 16:10:29 EST 2007

o	Bugfix: the timeout command always reported zero exit status
	after voluntary child process exit.

Mon Jun 26 18:44:17 EDT 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Fri Mar 17 09:25:57 EST 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Tue Jan  6 12:25:32 EST 2004

o       Workaround: mactime treated "," as a special character in
	file names, now it's 0xff.

o	Support for ext[23]fs inode file sizes > 32bits.

o	Bugfix: ils file sizes larger than unsigned long.

Tue Oct 14 18:23:49 EDT 2003

o       pcat now fully supports FreeBSD systems that do not have
	/proc mounted (it's no longer mounted by default with
	FreeBSD 5). pcat tries to use /proc if it can, and uses
	ptrace() and gropes kernel memory if it has to.  Operation
	without /proc requires super-user privileges.

Thu Oct  2 09:14:00 EDT 2003

o       Completed support for UFS1 and UFS2 in FreeBSD 5.x. However,
	the pcat command still requires that the /proc file system
	is mounted. This will be fixed later.

Sat Aug 30 19:21:29 EDT 2003

o	Preliminary FreeBSD 5.0 port. UFS2 support is still to be done.

Fri Aug  1 10:07:39 EDT 2003

o	RedHat 9.0 fix by Florin Andrei for the file command.

Sat Oct  5 13:47:29 EDT 2002

o	Restored tctutils compatibility, which broke with 20020916.

Mon Sep 16 14:39:09 EDT 2002

o       The lowest inode examined is now the lowest possible inode
	number instead of ROOTINUM.

Fri Mar 22 14:41:32 EST 2002

o	Added support for OpenBSD 3.0.

Wed Dec 12 13:21:00 EST 2001

o	tm_print() and tm_string() did unnecessary %HEX expansion
	of $ and ^ characters (Philippe Bourgeois).

o       mactime mis-sorted dates before/after early September 2001
	(Philippe Bourgeois, Brian Carrier).

Thu Oct  4 15:48:42 EDT 2001

o       Linux can handle files > 2Gbyte, but only if we specify
	extra compiler flags.

o	Solaris 8/64bit needs an extra include file in pcat.c.

Sun Sep  9 17:23:13 EDT 2001

o       Discovered to my horror that unrm on ext2fs can have an
	off-by-one error when testing a bit flag in the data block
	allocation bitmaps - and this problem is dependent on the
	Linux version being used.

o       Made the timeout command a little more verbose so it is
	less of a mystery when the child process is killed.

Thu Sep  6 13:51:00 EDT 2001

o	New grave-robber -I option for better control of icat/pcat.

Mon May 21 10:11:19 EDT 2001

o       findkey command-line error check was too conservative,
	and disallowed threshold of 256 different byte values
	(Ovanes Manucharyan).

Mon May 14 13:11:54 EDT 2001

o	Patch 06 broke lib/pass.cache.pl, which broke mactime.

o	Patch 06 did not update the patchlevel file.

Fri Mar 16 14:25:11 EST 2001

o	Fixed rpm root directory when running on a corpse.
	Marco Walter (SUN).

o	Fixed file(1) output parsing problem with multiple :
	characters.  Marco Walter (SUN)

Thu Feb 22 11:16:40 EST 2001

o	Patch by Brian Carrier (Purdue) so that fs_read_random()
	properly maintains file offset information.

Thu Feb  8 10:40:02 EST 2001

o	grave-robber dies silently when no group file is found.
	This bites when examining a partial machine corpse
	(bin/grave-robber). Klaus-Peter Kossakowski.

o	The "do first looks" code did not relocate to $TCT_HOME
	(lib/tree.pl). Klaus-Peter Kossakowski.

Tue Jan 23 18:44:36 EST 2001

o       The grave-robber did not correctly handle multi-line df
	output records (ps_spy.pl, suck_free_inodes.pl).

o       The grave-robber mounted corpse inode scan looked only at
	the corpse's "root" file system (suck_free_inodes.pl).

Mon Jan 22 11:23:32 EST 2001

o       grave-robber enabled the default -lPO flags when -c was
	specified, causing the program to gripe and exit.
	
Tue Sep 26 10:55:24 EDT 2000

o       Fixed a bug in fs_read_block() that caused icat etc. to
	remember the wrong file seek position, so that icat etc.
	would call lseek() too often. The other possibility, not
	calling lseek() when it should be called, did not happen
	due to an implementation artefact of the TCT utilities.
	Reported by Brian Carrier, Purdue University.

	Updated help-recovering-file document by Dan.

Sat Aug 12 10:51:16 EDT 2000

o       Added -V option to direct icat/ils/unrm for verbose output
	to stdout.  The -v option now sends verbose logging to stderr.

Tue Aug  8 11:46:19 PDT 2000

o	created "help-recovering-file", added note to README.FIRST

Sat Aug  5 11:34:26 PDT 2000

o	mactime time range repaired, fixed docs, -l option slightly
	fixed... perhaps deprecation in the future?

Fri Aug  4 20:06:15 EDT 2000

o	ils and ils2mac now include device names for more
	readable mactime reports of dead inodes.

o	mactime could not run with current working directory
	different from $TCT_HOME.

o	mactime -d could destroy an existing $DATA/body file.

o	mactime -d turned on verbosity, instead of letting -v do that.

o	mactime -s (suid) flag used an obsolete method of getting data.

o	ils2mac broke because perl split() can't see the difference
	between empty last fields and missing last fields.

o	mactime mistakenly skipped files with all-zero time stamps.

o	ils2mac now emits ls -l like info for mactime compatibility.

Tue Aug  1 13:50:20 PDT 2000

o	Prepend our directory list to @INC rather than clobbering it.

Mon Jul 31 17:37:35 PDT 2000

o	removed mention of sans.org from help-when-broken-into


src/fstools/ils.c
/*++
/* NAME
/*	ils 1
/* SUMMARY
/*	list inode information
/* SYNOPSIS
/* .ad
/* .fi
/*	\fBils\fR [\fB-eorvV\fR] [\fB-f \fIfstype\fR]
/*		\fIdevice\fR [\fIstart-stop\fR ...]
/*
/*	\fBils\fR [\fB-aAlLvVzZ\fR] [\fB-f \fIfstype\fR]
/*		\fIdevice\fR [\fIstart-stop\fR ...]
/* DESCRIPTION
/*	\fBils\fR opens the named \fIdevice\fR and lists inode information.
/*	By default, \fBils\fR lists only the inodes of removed files.
/*
/* 	Arguments:
/* .IP \fB-e\fR
/*	List every inode in the file system.
/* .IP "\fB-f\fI fstype\fR"
/*	Specifies the file system type. The default file system type
/*	is system dependent. With most UNIX systems the default type
/*	is \fBffs\fR (Berkeley fast file system). With Linux the default
/*	type is \fBext2fs\fR (second extended file system).
/* .IP \fB-o\fR
/*	List only inodes of removed files that are still open or executing.
/*	This option is short-hand notation for \fB-aL\fR
/*	(see the \fBfine controls\fR section below).
/* .IP \fB-r\fR
/*	List only inodes of removed files. This option is short-hand notation
/*	for \fB-LZ\fR
/*	(see the \fBfine controls\fR section below).
/* .IP \fB-v\fR
/*	Turn on verbose mode, output to stderr.
/* .IP \fB-V\fR
/*	Turn on verbose mode, output to stdout.
/* .IP \fIdevice\fR
/*	Disk special file, or regular file containing a disk image.
/*	On UNIX systems, raw mode disk access may give better performance
/*	than block mode disk access.  LINUX disk device drivers support
/*	only block mode disk access.
/* .IP "\fIstart-stop\fR ..."
/*	Examine the specified inode number or number range. Either the
/*	\fIstart\fR, the \fIstop\fR, or the \fI-stop\fR may be omitted.
/* .PP
/*	Fine controls:
/* .IP \fB-a\fR
/*	List only allocated inodes: these belong to files with at least one
/*	directory entry in the file system, and to removed files that
/*	are still open or executing.
/* .IP \fB-A\fR
/*	List only unallocated inodes: these belong to files that no longer
/*	exist.
/* .IP \fB-l\fR
/*	List only inodes with at least one hard link. These belong to files
/*	with at least one directory entry in the file system.
/* .IP \fB-L\fR
/*	List only inodes without any hard links. These belong to files that no
/*	longer exist, and to removed files that are still open or executing.
/* .IP \fB-z\fR
/*	List only inodes with zero status change time. Presumably, these
/*	inodes were never used.
/* .IP \fB-Z\fR
/*	List only inodes with non-zero status change time. Presumably, these
/*	belong to files that still exist, or that existed in the past.
/* .PP
/*	The output format is in time machine format, as described in
/*	tm-format(5). The output begins with a two-line header that
/*	describes the data origin, and is followed by a one-line header
/*	that lists the names of the data attributes that make up the
/*	remainder of the output:
/* .IP \fBst_ino\fR
/*	The inode number.
/* .IP \fBst_alloc\fR
/*	Allocation status: `a' for allocated inode, `f' for free inode.
/* .IP \fBst_uid\fR
/*	Owner user ID.
/* .IP \fBst_gid\fR
/*	Owner group ID.
/* .IP \fBst_mtime\fR
/*	UNIX time (seconds) of last file modification.
/* .IP \fBst_atime\fR
/*	UNIX time (seconds) of last file access.
/* .IP \fBst_ctime\fR
/*	UNIX time (seconds) of last inode status change.
/* .IP \fBst_dtime\fR
/*	UNIX time (seconds) of file deletion (LINUX only).
/* .IP \fBst_mode\fR
/*	File type and permissions (octal).
/* .IP \fBst_nlink\fR
/*	Number of hard links.
/* .IP \fBst_size\fR
/*	File size in bytes.
/* .IP \fBst_block0,st_block1\fR
/*	The first two entries in the direct block address list.
/* SEE ALSO
/*	mactime(1), mtime, atime, ctime reporter
/*	tm-format(5), time machine data format
/* BUGS
/*	\fBils\fR should support more file system types. Right now, support
/*	is limited to \fBext2fs\fR when built on Linux, and \fBffs\fR when
/*	built on Solaris and BSD systems.
/* LICENSE
/*	This software is distributed under the IBM Public License.
/* AUTHOR(S)
/*	Wietse Venema
/*	IBM T.J. Watson Research
/*	P.O. Box 704
/*	Yorktown Heights, NY 10598, USA
/*--*/

#include "fs_tools.h"
#include "error.h"
#include "split_at.h"

FILE   *logfp;

#define DEF_FLAGS	(FS_FLAG_USED | FS_FLAG_UNLINK)

/* atoinum - convert string to inode number */

INUM_T  atoinum(const char *str)
{
    char   *cp;
    INUM_T  inum;

    if (*str == 0)
	return (0);
    inum = STRTOUL(str, &cp, 0);
    if (*cp || cp == str)
	error("bad inode number: %s", str);
    return (inum);
}

/* usage - explain and terminate */

static void usage()
{
    error("usage: %s [-e (everything)] [-f fstype] [-o (removed but still open)] [-r (removed)] [-vV] device [inum... ]",
	  progname);
}

/* print_header - print time machine header */

static void print_header(const char *device)
{
    char    hostnamebuf[BUFSIZ];
    unsigned long now;

    if (gethostname(hostnamebuf, sizeof(hostnamebuf) - 1) < 0)
	error("gethostname: %m");
    hostnamebuf[sizeof(hostnamebuf) - 1] = 0;
    now = time((time_t *) 0);

    /*
     * Identify table type and table origin.
     */
    printf("class|host|device|start_time\n");
    printf("ils|%s|%s|%lu\n", hostnamebuf, device, now);

    /*
     * Identify the fields in the data that follow.
     */
    printf("st_ino|st_alloc|st_uid|st_gid|st_mtime|st_atime|st_ctime");
#ifdef HAVE_DTIME
    printf("|st_dtime");
#endif
    printf("|st_mode|st_nlink|st_size|st_block0|st_block1\n");
}

/* print_inode - list generic inode */

static void print_inode(INUM_T inum, FS_INODE *fs_inode, int flags,
			        char *unused_context)
{
    printf("%lu|%c|%d|%d|%lu|%lu|%lu",
	   (ULONG) inum, (flags & FS_FLAG_ALLOC) ? 'a' : 'f',
	   (int) fs_inode->uid, (int) fs_inode->gid,
	   (ULONG) fs_inode->mtime, (ULONG) fs_inode->atime,
	   (ULONG) fs_inode->ctime);
#ifdef HAVE_DTIME
    printf("|%lu", (ULONG) fs_inode->dtime);
#endif
    if (sizeof(fs_inode->size) <= sizeof(unsigned long))
	printf("|%lo|%d|%lu|%lu|%lu\n",
	       (ULONG) fs_inode->mode, (int) fs_inode->nlink,
	       (ULONG) fs_inode->size, (ULONG) fs_inode->direct_addr[0],
	       (ULONG) fs_inode->direct_addr[1]);
    else
	printf("|%lo|%d|%llu|%lu|%lu\n",
	       (ULONG) fs_inode->mode, (int) fs_inode->nlink,
	       (unsigned long long) fs_inode->size, (ULONG) fs_inode->direct_addr[0],
	       (ULONG) fs_inode->direct_addr[1]);
}

/* main - open file system, list inode info */

int     main(int argc, char **argv)
{
    FS_INFO *fs;
    char   *start;
    char   *last;
    INUM_T  istart;
    INUM_T  ilast;
    int     ch;
    int     flags = 0;
    char   *fstype = DEF_FSTYPE;

    progname = argv[0];

    /*
     * Provide convenience options for the most commonly selected feature
     * combinations.
     */
    while ((ch = getopt(argc, argv, "aAef:lLorvVzZ")) > 0) {
	switch (ch) {
	default:
	    usage();
	case 'e':
	    flags |= ~0;
	    break;
	case 'f':
	    fstype = optarg;
	    break;
	case 'o':
	    flags |= (FS_FLAG_ALLOC | FS_FLAG_UNLINK);
	    break;
	case 'r':
	    flags |= (FS_FLAG_UNLINK | FS_FLAG_USED);
	    break;
	case 'v':
	    verbose++;
	    logfp = stderr;
	    break;
	case 'V':
	    verbose++;
	    logfp = stdout;
	    break;

	    /*
	     * Provide fine controls to tweak one feature at a time.
	     */
	case 'a':
	    flags |= FS_FLAG_ALLOC;
	    flags &= ~FS_FLAG_UNALLOC;
	    break;
	case 'A':
	    flags |= FS_FLAG_UNALLOC;
	    flags &= ~FS_FLAG_ALLOC;
	    break;
	case 'l':
	    flags |= FS_FLAG_LINK;
	    flags &= ~FS_FLAG_UNLINK;
	    break;
	case 'L':
	    flags |= FS_FLAG_UNLINK;
	    flags &= ~FS_FLAG_LINK;
	    break;
	case 'z':
	    flags |= FS_FLAG_UNUSED;
	    flags &= ~FS_FLAG_USED;
	    break;
	case 'Z':
	    flags |= FS_FLAG_USED;
	    flags &= ~FS_FLAG_UNUSED;
	    break;
	}
    }

    if (optind >= argc)
	usage();

    /*
     * Apply rules for default settings. Assume a "don't care" condition when
     * nothing is explicitly selected from a specific feature category.
     */
    if (flags == 0)
	flags = DEF_FLAGS;
    if ((flags & (FS_FLAG_ALLOC | FS_FLAG_UNALLOC)) == 0)
	flags |= FS_FLAG_ALLOC | FS_FLAG_UNALLOC;
    if ((flags & (FS_FLAG_LINK | FS_FLAG_UNLINK)) == 0)
	flags |= FS_FLAG_LINK | FS_FLAG_UNLINK;
    if ((flags & (FS_FLAG_USED | FS_FLAG_UNUSED)) == 0)
	flags |= FS_FLAG_USED | FS_FLAG_UNUSED;

    /*
     * Open the file system.
     */
    fs = fs_open(argv[optind], fstype);

    /*
     * Print the time machine header.
     */
    print_header(argv[optind]);

    /*
     * List the named inodes.
     */
    optind++;
    if (optind < argc) {
	while ((start = argv[optind]) != 0) {
	    last = split_at(start, '-');
	    istart = (*start ? atoinum(start) : fs->start_inum);
	    ilast = (!last ? istart : *last ? atoinum(last) : fs->last_inum);
	    fs->inode_walk(fs, istart, ilast, flags, print_inode, (char *) 0);
	    optind++;
	}
    }

    /*
     * List all inodes, subject to the specified restrictions.
     */
    else {
	fs->inode_walk(fs, fs->start_inum, fs->last_inum,
		       flags, print_inode, (char *) 0);
    }
    fs->close(fs);
    exit(0);
}
